README.md
<a href="https://github.com/w3f/Open-Grants-Program/pull/268"><img src="https://user-images.githubusercontent.com/1693789/156277834-ed433b60-9e82-4267-8b4f-e30438dbec54.png" alt="oak-web3-open-grant" /></a>
OAK(Onchain Autonomous Framework) is a unique blockchain built on Substrate framework with event-driven smart contract VM, autonomous transactions, and on-chain scheduler.

Documentation
----------

* [Website](https://oak.tech/)
* [Documentation](https://docs.oak.tech/)

Community
--------- 

* General discussion: [Telegram](https://t.me/OAK_Announcements)
* Technical discussion: [Discord](https://discord.gg/7W9UDvsbwh)
* Subscribe on [OAK Twitter](https://twitter.com/oak_network)
* Subscribe on [Founder's Twitter](https://twitter.com/chrisli2046)

Introduction
============

**OAK, or Onchain Automation Framework, is equipped with a novel smart contract virtual machine that supports an event-driven execution model, enabling developers to build fully autonomous decentralized application.** By extending the current set of atomic operations, namely, opcodes of EVM, OAK introduces an innovative way for contracts to interact with each other. Contracts can emit signal events, on which other contracts can listen. Once an event is triggered, corresponding handler functions are automatically executed as a new type of transaction, signal transaction. Applications implemented with the new approach will eliminate the dependency of unreliable mechanisms like off-chain relay servers, and in return, to significantly simplify the execution flow of the application and can avoid security risks such as relay manipulation attacks.

Based on the above, OAK has some features.
- **OAK Virtual Machine**
- **Autonomous Transactions**
- **On-chain Relayer**
- **Collator Staking**

Live Networks
============

- `neumann`: testnet parachain (January 2022)
- `turing-staging`: rococo parachain (May 2022)
- `turing`: kusama parachain (April 2022)
- `oak`: polkadot parachain (coming soon)

Install OAK Blockchain 
=============

* OAK releases [releases](https://github.com/OAK-Foundation/OAK-blockchain/releases).

Building from source
--------------------

Ensure you have Rust and the support software (see shell.nix for the latest functional toolchain):

    curl https://sh.rustup.rs -sSf | sh
    # on Windows download and run rustup-init.exe
    # from https://rustup.rs instead

    rustup update nightly
    rustup target add wasm32-unknown-unknown --toolchain nightly

You will also need to install the following dependencies:

* Linux: `sudo apt install cmake git clang libclang-dev build-essential`
* Mac: `brew install cmake git llvm`
* Windows: Download and install the Pre Build Windows binaries of LLVM from http://releases.llvm.org/download.html

Install additional build tools:

    cargo +nightly install --git https://github.com/alexcrichton/wasm-gc

Install the OAK node from git source:

    git clone git@github.com:OAK-Foundation/OAK-blockchain.git    

Build your executable:
```bash
    cargo build --release --features neumann-node
    # OR
    cargo build --release --features turing-node
    # OR
    cargo build --release --features oak-node
```

Run your Local Network
-----------

Launch a local setup including a Relay Chain and a Parachain.
Note: local PARA_ID is defaulted to 2000

### Launch the Relay Chain

```bash
# Compile Polkadot with the real overseer feature
git clone https://github.com/paritytech/polkadot
cargo build --release

# Alice
./target/release/polkadot \
--alice \
--validator \
--tmp \
--chain ../OAK-blockchain/resources/rococo-local.json \
--port 30333 \
--ws-port 9944

# Bob (In a separate terminal)
./target/release/polkadot \
--bob \
--validator \
--tmp \
--chain ../OAK-blockchain/resources/rococo-local.json \
--port 30334 \
--ws-port 9945
```

### Reserve the Parachain slot

1. Navigate to [Local relay parachain screen](https://polkadot.js.org/apps/?rpc=ws%3A%2F%2F127.0.0.1%3A9944#/parachains/parathreads)
2. Click `+ ParaId`
3. Reserve `paraid` with the following paramaters
    - `reserve from`: `Alice`
    - `parachain id`: 2000
    - `reserved deposit`: <whatever the default is>


### Prep the Parachain

```bash
# Compile
git clone https://github.com/OAK-Foundation/OAK-blockchain
cargo build --release --features neumann-node

# Export genesis state
./target/release/oak-collator export-genesis-state > genesis-state

# Export genesis wasm
./target/release/oak-collator export-genesis-wasm > genesis-wasm

# Collator1
./target/release/oak-collator \
--alice \
--collator \
--force-authoring \
--tmp \
--port 40333 \
--ws-port 9946 \
-- \
--execution wasm \
--chain resources/rococo-local.json \
--port 30335 \
--ws-port 9977 
```

To make testing locally easier use the command below.  The `dev-queue` flag will allow for putting a task directly on the task queue as opposed to waiting until the next hour to schedule a task.  This works when the `execution_times` passed to schedule a task equals `[0]`.
```
cargo build --release --features neumann-node --features dev-queue
```

### Register the parachain

1. Navigate to [Local relay sudo extrinsic](https://polkadot.js.org/apps/?rpc=ws%3A%2F%2F127.0.0.1%3A9944#/sudo)
2. Register your local parachain with the local relay chain (see the image below for the extrinsic). 
Note** the files mentioned are the ones you generated above.

![image](https://user-images.githubusercontent.com/2915325/99548884-1be13580-2987-11eb-9a8b-20be658d34f9.png)


### Test the parachain

1. Navigate to [Local parachain](https://polkadot.js.org/apps/?rpc=ws%3A%2F%2F127.0.0.1%3A9946#/explorer)
2. WAIT.... It takes a while for the registration process to finish. 

Future Work
------------
Here are the key milestones.

1. Start the crowdloan with Kusama network
2. Become a Kusama Parachain (TBA)
3. Become a Polkadot Parachain. (TBA)

If you have any questions, please ask us on [Discord](https://discord.gg/7W9UDvsbwh)

Contacts
--------

**Maintainers**

* [OAK Development Team](https://github.com/orgs/OAK-Foundation/people)

* * *

OAK blockchain is licensed under the GPLv3.0 by OAK Foundation.


CHANGELOG.md
# CHANGELOG -- DEPRECATED
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

# Deprecated in favor of [Releases](https://github.com/OAK-Foundation/OAK-blockchain/releases)

## [1.7.0] - 2022-08-18
- Adding try-runtime to Turing
- Adding Currencies, PolkadotXcm, and XTokens to Turing closed call filter (Valve) 
- Add xcmp-handler pallet to execute xcm, send xcm, and provide xcm fees
- Add pallet-utility
- Schedule XCM through automation-time extrinsic
- Price triggers (Alpha)

## [1.6.0] - 2022-08-10
- Update to substrate 0.9.26
- Added base XCMP-handler pallet
- Updated XCMP-handler pallet to store chain/currency information
- Updated XCMP-handler pallet to create XCM instructions for a Transact with OAK tokens
- RPC to get accountId for XCM Descend + Transact
- RPC to get fees for scheduling an XCM task
- Update auto-compounding to work with new parachain-staking locks
- Removed tight coupling between valve and automation-time pallet

## [1.5.0] - 2022-07-19
- RPC to get execution fee for automation time
- New auto re-staking automation time feature
- RPC to get ideal frequency for re-staking
- Added proxy pallet
- Added multisig pallet

## [283] - 2022-06-23
Note: Runtime-only release
- Update Neumann and Turing governance parameters

## [1.4.0] - 2022-06-01
- Fixed execution fees for automation time
- Update to substrate 0.9.20
- Added Parallel Heiko tokens
- Changed KUSD to AUSD
- Updating fee burn rate to 20%
- Provide on-chain identity registration

## [1.3.0] - 2022-05-17
- Moving Turing xcm configs to their own file.
- Allow foreign tokens to get stored on-chain
- Enable cross chain token transfers
- Lower existential deposit from .1 to .01
- Adding Karura tokens
- Ensure CollatorRegistration before joining candidate pool
- Update to substrate 0.9.19

## [279] - 2022-04-29
- Change scheduled time slots from minutes to hours. Migration used to clear all tasks in existing maps and queues.
- Change max tasks in a given slot to 256 per hour.
- Added recurring tasks feature to schedule up to 24 recurring executions for a single task.

## [1.2.8] - 2022-04-05
- Updated to substrate 0.9.18

## [277] - 2022-03-14
Note: Runtime-only release
- Disable treasury burn in runtime
- Split update queue behavior for missed and scheduled queues
- Added Democracy pallet and Technical Committee (#bump-tx-version)
- Added Turing initial allocation & vesting for wallets
- Change vesting from the 5th of the month to the 1st of the month

## [1.2.7] - 2022-03-10
- Removed Quadratic funding pallet; can now be found [here](https://github.com/OAK-Foundation/quadratic-funding-pallet)
- Added inclusion fees for OAK transactions
- Modified execution fees
- Added weights for pallet-valve
- Upgraded polkadot client to v0.9.17
- Added vesting on initialize
- Benchmarks for automation time pallet
- Safe math calculations

## [1.2.6] - 2022-03-01
- Add Turing runtime and split chain-specific code (#chain-fork)
- Add telemetry for node infra
- Default Turing chain_spec on genesis has all gates closed
- Adding pallet-vesting for OAK's own vesting schedule and distribution

## [1.2.5] - 2022-02-20
- Automation Time RPC called automationTime_generateTaskId
- Added onfinality neumann cli script
- Missed tasks for a given time window will not be run
- Action: wallet to wallet transfer
- Neumann tasks cannot be scheduled farther than 1 week from today
- Change ss58 prefix to 51 for NEU, TUR, OAK (#chain-fork)

## [1.2.4] - 2022-02-08
### Added
- automationTime pallet deployed for a future events (N=1)
- parts of the valve pallet implemented and deployed


node/src/command.rs
use std::net::SocketAddr;

use codec::Encode;
use cumulus_client_cli::generate_genesis_block;
use cumulus_primitives_core::ParaId;
use frame_benchmarking_cli::{BenchmarkCmd, SUBSTRATE_REFERENCE_HARDWARE};
use log::info;
use sc_cli::{
	ChainSpec, CliConfiguration, DefaultConfigurationValues, ImportParams, KeystoreParams,
	NetworkParams, Result, RuntimeVersion, SharedParams, SubstrateCli,
};
use sc_service::config::{BasePath, PrometheusConfig};
use sp_core::hexdisplay::HexDisplay;
use sp_runtime::traits::{AccountIdConversion, Block as BlockT};

use crate::{
	chain_spec::{self, IdentifyVariant},
	cli::{Cli, RelayChainCli, Subcommand},
	service,
};

fn load_spec(id: &str) -> std::result::Result<Box<dyn ChainSpec>, String> {
	Ok(match id {
		#[cfg(feature = "neumann-node")]
		"" | "dev" => Box::new(chain_spec::neumann::development_config()),
		#[cfg(feature = "turing-node")]
		"turing-dev" => Box::new(chain_spec::turing::turing_development_config()),
		#[cfg(feature = "turing-node")]
		"turing-staging" => Box::new(chain_spec::turing::turing_staging()?),
		#[cfg(feature = "turing-node")]
		"turing" => Box::new(chain_spec::turing::turing_live()?),
		#[cfg(feature = "oak-node")]
		"oak-dev" => Box::new(chain_spec::oak::oak_development_config()),
		#[cfg(feature = "oak-node")]
		"oak-staging" => Box::new(chain_spec::oak::oak_staging()),
		#[cfg(feature = "oak-node")]
		"oak" => Box::new(chain_spec::oak::oak_live()),
		path => {
			let path = std::path::PathBuf::from(path);
			let chain_spec = Box::new(chain_spec::DummyChainSpec::from_json_file(path.clone())?)
				as Box<dyn sc_service::ChainSpec>;

			match chain_spec {
				chain_spec if chain_spec.is_turing() => {
					#[cfg(feature = "turing-node")]
					{
						Box::new(chain_spec::turing::ChainSpec::from_json_file(path)?)
					}
					#[cfg(not(feature = "turing-node"))]
					return Err(service::TURING_RUNTIME_NOT_AVAILABLE.into())
				},
				chain_spec if chain_spec.is_oak() => {
					#[cfg(feature = "oak-node")]
					{
						Box::new(chain_spec::oak::ChainSpec::from_json_file(path)?)
					}
					#[cfg(not(feature = "oak-node"))]
					return Err(service::OAK_RUNTIME_NOT_AVAILABLE.into())
				},
				_ => {
					#[cfg(feature = "neumann-node")]
					{
						Box::new(chain_spec::neumann::ChainSpec::from_json_file(path)?)
					}
					#[cfg(not(feature = "neumann-node"))]
					return Err(service::NEUMANN_RUNTIME_NOT_AVAILABLE.into())
				},
			}
		},
	})
}

impl SubstrateCli for Cli {
	fn impl_name() -> String {
		"OAK Collator".into()
	}

	fn impl_version() -> String {
		env!("SUBSTRATE_CLI_IMPL_VERSION").into()
	}

	fn description() -> String {
		format!(
			"OAK Collator\n\nThe command-line arguments provided first will be \
			passed to the parachain node, while the arguments provided after -- will be passed \
			to the relay chain node.\n\n\
			{} <parachain-args> -- <relay-chain-args>",
			Self::executable_name()
		)
	}

	fn author() -> String {
		env!("CARGO_PKG_AUTHORS").into()
	}

	fn support_url() -> String {
		"https://github.com/OAK-Foundation/OAK-blockchain/issues/new".into()
	}

	fn copyright_start_year() -> i32 {
		2021
	}

	fn load_spec(&self, id: &str) -> std::result::Result<Box<dyn sc_service::ChainSpec>, String> {
		load_spec(id)
	}

	fn native_runtime_version(
		chain_spec: &Box<dyn sc_service::ChainSpec>,
	) -> &'static RuntimeVersion {
		match chain_spec {
			chain_spec if chain_spec.is_turing() => {
				#[cfg(not(feature = "turing-node"))]
				panic!("{}", service::TURING_RUNTIME_NOT_AVAILABLE);

				#[cfg(feature = "turing-node")]
				return &service::turing_runtime::VERSION
			},
			chain_spec if chain_spec.is_oak() => {
				#[cfg(not(feature = "oak-node"))]
				panic!("{}", service::OAK_RUNTIME_NOT_AVAILABLE);

				#[cfg(feature = "oak-node")]
				return &service::oak_runtime::VERSION
			},
			_ => {
				#[cfg(not(feature = "neumann-node"))]
				panic!("{}", service::NEUMANN_RUNTIME_NOT_AVAILABLE);

				#[cfg(feature = "neumann-node")]
				return &service::neumann_runtime::VERSION
			},
		}
	}
}

impl SubstrateCli for RelayChainCli {
	fn impl_name() -> String {
		"OAK Collator".into()
	}

	fn impl_version() -> String {
		env!("SUBSTRATE_CLI_IMPL_VERSION").into()
	}

	fn description() -> String {
		format!(
			"OAK Collator\n\nThe command-line arguments provided first will be \
			passed to the parachain node, while the arguments provided after -- will be passed \
			to the relay chain node.\n\n\
			{} <parachain-args> -- <relay-chain-args>",
			Self::executable_name()
		)
	}

	fn author() -> String {
		env!("CARGO_PKG_AUTHORS").into()
	}

	fn support_url() -> String {
		"https://github.com/OAK-Foundation/OAK-blockchain/issues/new".into()
	}

	fn copyright_start_year() -> i32 {
		2021
	}

	fn load_spec(&self, id: &str) -> std::result::Result<Box<dyn sc_service::ChainSpec>, String> {
		polkadot_cli::Cli::from_iter([RelayChainCli::executable_name()].iter()).load_spec(id)
	}

	fn native_runtime_version(
		chain_spec: &Box<dyn sc_service::ChainSpec>,
	) -> &'static RuntimeVersion {
		polkadot_cli::Cli::native_runtime_version(chain_spec)
	}
}

macro_rules! construct_async_run {
	(|$components:ident, $cli:ident, $cmd:ident, $config:ident| $( $code:tt )* ) => {{
		let runner = $cli.create_runner($cmd)?;
		let chain_spec = &runner.config().chain_spec;

		with_runtime_or_err!(chain_spec, {
			{
				runner.async_run(|$config| {
					let $components = service::new_partial::<
						RuntimeApi,
						Executor,
						_
					>(
						&$config,
						crate::service::parachain_build_import_queue,
					)?;
					let task_manager = $components.task_manager;
					{ $( $code )* }.map(|v| (v, task_manager))
				})
			}
		})
	}}
}

macro_rules! with_runtime_or_err {
	($chain_spec:expr, { $( $code:tt )* }) => {
		match $chain_spec {
			chain_spec if chain_spec.is_turing() => {
				#[cfg(feature = "turing-node")]
				#[allow(unused_imports)]
				use service::{turing_runtime::{Block, RuntimeApi}, TuringExecutor as Executor};
				#[cfg(feature = "turing-node")]
				$( $code )*

				#[cfg(not(feature = "turing-node"))]
				return Err(service::TURING_RUNTIME_NOT_AVAILABLE.into());
			},
			chain_spec if chain_spec.is_oak() => {
				#[cfg(feature = "oak-node")]
				#[allow(unused_imports)]
				use service::{oak_runtime::{Block, RuntimeApi}, OakExecutor as Executor};
				#[cfg(feature = "oak-node")]
				$( $code )*

				#[cfg(not(feature = "oak-node"))]
				return Err(service::OAK_RUNTIME_NOT_AVAILABLE.into());
			},
			_ => {
				#[cfg(feature = "neumann-node")]
				#[allow(unused_imports)]
				use service::{neumann_runtime::{Block, RuntimeApi}, NeumannExecutor as Executor};
				#[cfg(feature = "neumann-node")]
				$( $code )*

				#[cfg(not(feature = "neumann-node"))]
				return Err(service::NEUMANN_RUNTIME_NOT_AVAILABLE.into());
			},
		}
	}
}

/// Parse command line arguments into service configuration.
pub fn run() -> Result<()> {
	let cli = Cli::from_args();

	match &cli.subcommand {
		Some(Subcommand::BuildSpec(cmd)) => {
			let runner = cli.create_runner(cmd)?;
			runner.sync_run(|config| cmd.run(config.chain_spec, config.network))
		},
		Some(Subcommand::CheckBlock(cmd)) => {
			construct_async_run!(|components, cli, cmd, config| {
				Ok(cmd.run(components.client, components.import_queue))
			})
		},
		Some(Subcommand::ExportBlocks(cmd)) => {
			construct_async_run!(|components, cli, cmd, config| {
				Ok(cmd.run(components.client, config.database))
			})
		},
		Some(Subcommand::ExportState(cmd)) => {
			construct_async_run!(|components, cli, cmd, config| {
				Ok(cmd.run(components.client, config.chain_spec))
			})
		},
		Some(Subcommand::ImportBlocks(cmd)) => {
			construct_async_run!(|components, cli, cmd, config| {
				Ok(cmd.run(components.client, components.import_queue))
			})
		},
		Some(Subcommand::Revert(cmd)) => {
			construct_async_run!(|components, cli, cmd, config| {
				Ok(cmd.run(components.client, components.backend, None))
			})
		},
		Some(Subcommand::PurgeChain(cmd)) => {
			let runner = cli.create_runner(cmd)?;

			runner.sync_run(|config| {
				let polkadot_cli = RelayChainCli::new(
					&config,
					[RelayChainCli::executable_name()].iter().chain(cli.relay_chain_args.iter()),
				);

				let polkadot_config = SubstrateCli::create_configuration(
					&polkadot_cli,
					&polkadot_cli,
					config.tokio_handle.clone(),
				)
				.map_err(|err| format!("Relay chain argument error: {}", err))?;

				cmd.run(config, polkadot_config)
			})
		},
		Some(Subcommand::ExportGenesisState(cmd)) => {
			let runner = cli.create_runner(cmd)?;
			let chain_spec = &runner.config().chain_spec;
			with_runtime_or_err!(chain_spec, {
				{
					runner.sync_run(|_config| {
						let spec =
							cli.load_spec(&cmd.shared_params.chain.clone().unwrap_or_default())?;
						let state_version = Cli::native_runtime_version(&spec).state_version();
						cmd.run::<Block>(&*spec, state_version)
					})
				}
			})
		},
		Some(Subcommand::ExportGenesisWasm(cmd)) => {
			let runner = cli.create_runner(cmd)?;
			runner.sync_run(|_config| {
				let spec = cli.load_spec(&cmd.shared_params.chain.clone().unwrap_or_default())?;
				cmd.run(&*spec)
			})
		},
		Some(Subcommand::Benchmark(cmd)) => {
			let runner = cli.create_runner(cmd)?;
			let chain_spec = &runner.config().chain_spec;
			// Switch on the concrete benchmark sub-command-
			with_runtime_or_err!(chain_spec, {
				{
					match cmd {
						BenchmarkCmd::Pallet(cmd) =>
							if cfg!(feature = "runtime-benchmarks") {
								runner.sync_run(|config| cmd.run::<Block, Executor>(config))
							} else {
								Err("Benchmarking wasn't enabled when building the node. \
					You can enable it with `--features runtime-benchmarks`."
									.into())
							},
						BenchmarkCmd::Block(cmd) => runner.sync_run(|config| {
							let partials = service::new_partial::<RuntimeApi, Executor, _>(
								&config,
								crate::service::parachain_build_import_queue,
							)?;
							cmd.run(partials.client)
						}),
						BenchmarkCmd::Storage(cmd) => runner.sync_run(|config| {
							let partials = service::new_partial::<RuntimeApi, Executor, _>(
								&config,
								crate::service::parachain_build_import_queue,
							)?;
							let db = partials.backend.expose_db();
							let storage = partials.backend.expose_storage();

							cmd.run(config, partials.client.clone(), db, storage)
						}),
						BenchmarkCmd::Overhead(_) => Err("Unsupported benchmarking command".into()),
						BenchmarkCmd::Machine(cmd) => runner.sync_run(|config| {
							cmd.run(&config, SUBSTRATE_REFERENCE_HARDWARE.clone())
						}),
					}
				}
			})
		},
		#[cfg(feature = "try-runtime")]
		Some(Subcommand::TryRuntime(cmd)) => {
			let runner = cli.create_runner(cmd)?;
			let chain_spec = &runner.config().chain_spec;
			with_runtime_or_err!(chain_spec, {
				{
					// grab the task manager.
					let registry =
						&runner.config().prometheus_config.as_ref().map(|cfg| &cfg.registry);
					let task_manager = sc_service::TaskManager::new(
						runner.config().tokio_handle.clone(),
						*registry,
					)
					.map_err(|e| format!("Error: {:?}", e))?;

					runner
						.async_run(|config| Ok((cmd.run::<Block, Executor>(config), task_manager)))
				}
			})
		},
		None => {
			let runner = cli.create_runner(&cli.run.normalize())?;
			let collator_options = cli.run.collator_options();

			runner.run_node_until_exit(|config| async move {
				let hwbench = if !cli.no_hardware_benchmarks {
					config.database.path().map(|database_path| {
						let _ = std::fs::create_dir_all(&database_path);
						sc_sysinfo::gather_hwbench(Some(database_path))
					})
				} else {
					None
				};

				let chain_spec = &config.chain_spec;
				let para_id = chain_spec::Extensions::try_get(&*config.chain_spec)
					.map(|e| e.para_id)
					.ok_or_else(|| "Could not find parachain ID in chain-spec.")?;

				let polkadot_cli = RelayChainCli::new(
					&config,
					[RelayChainCli::executable_name()].iter().chain(cli.relay_chain_args.iter()),
				);

				let id = ParaId::from(para_id);
				let parachain_account =
					AccountIdConversion::<polkadot_primitives::v2::AccountId>::into_account_truncating(&id);
				let tokio_handle = config.tokio_handle.clone();
				let polkadot_config =
					SubstrateCli::create_configuration(&polkadot_cli, &polkadot_cli, tokio_handle)
						.map_err(|err| format!("Relay chain argument error: {}", err))?;

				let state_version = Cli::native_runtime_version(&config.chain_spec).state_version();

				let genesis_state = with_runtime_or_err!(chain_spec, {
					{
						let block: Block = generate_genesis_block(&**chain_spec, state_version)?;
						format!("0x{:?}", HexDisplay::from(&block.header().encode()))
					}
				});

				info!("Parachain id: {:?}", id);
				info!("Parachain Account: {}", parachain_account);
				info!("Parachain genesis state: {:?}", genesis_state);
				info!("Is collating: {}", if config.role.is_authority() { "yes" } else { "no" });

				with_runtime_or_err!(chain_spec, {
					{
						crate::service::start_parachain_node::<RuntimeApi, Executor>(
							config,
							polkadot_config,
							collator_options,
							id,
							hwbench,
						)
						.await
						.map(|r| r.0)
						.map_err(Into::into)
					}
				})
			})
		},
	}
}

impl DefaultConfigurationValues for RelayChainCli {
	fn p2p_listen_port() -> u16 {
		30334
	}

	fn rpc_ws_listen_port() -> u16 {
		9945
	}

	fn rpc_http_listen_port() -> u16 {
		9934
	}

	fn prometheus_listen_port() -> u16 {
		9616
	}
}

impl CliConfiguration<Self> for RelayChainCli {
	fn shared_params(&self) -> &SharedParams {
		self.base.base.shared_params()
	}

	fn import_params(&self) -> Option<&ImportParams> {
		self.base.base.import_params()
	}

	fn network_params(&self) -> Option<&NetworkParams> {
		self.base.base.network_params()
	}

	fn keystore_params(&self) -> Option<&KeystoreParams> {
		self.base.base.keystore_params()
	}

	fn base_path(&self) -> Result<Option<BasePath>> {
		Ok(self
			.shared_params()
			.base_path()
			.or_else(|| self.base_path.clone().map(Into::into)))
	}

	fn rpc_http(&self, default_listen_port: u16) -> Result<Option<SocketAddr>> {
		self.base.base.rpc_http(default_listen_port)
	}

	fn rpc_ipc(&self) -> Result<Option<String>> {
		self.base.base.rpc_ipc()
	}

	fn rpc_ws(&self, default_listen_port: u16) -> Result<Option<SocketAddr>> {
		self.base.base.rpc_ws(default_listen_port)
	}

	fn prometheus_config(
		&self,
		default_listen_port: u16,
		chain_spec: &Box<dyn sc_service::ChainSpec>,
	) -> Result<Option<PrometheusConfig>> {
		self.base.base.prometheus_config(default_listen_port, chain_spec)
	}

	fn init<F>(
		&self,
		_support_url: &String,
		_impl_version: &String,
		_logger_hook: F,
		_config: &sc_service::Configuration,
	) -> Result<()>
	where
		F: FnOnce(&mut sc_cli::LoggerBuilder, &sc_service::Configuration),
	{
		unreachable!("PolkadotCli is never initialized; qed");
	}

	fn chain_id(&self, is_dev: bool) -> Result<String> {
		let chain_id = self.base.base.chain_id(is_dev)?;

		Ok(if chain_id.is_empty() { self.chain_id.clone().unwrap_or_default() } else { chain_id })
	}

	fn role(&self, is_dev: bool) -> Result<sc_service::Role> {
		self.base.base.role(is_dev)
	}

	fn transaction_pool(&self) -> Result<sc_service::config::TransactionPoolOptions> {
		self.base.base.transaction_pool()
	}

	fn state_cache_child_ratio(&self) -> Result<Option<usize>> {
		self.base.base.state_cache_child_ratio()
	}

	fn rpc_methods(&self) -> Result<sc_service::config::RpcMethods> {
		self.base.base.rpc_methods()
	}

	fn rpc_ws_max_connections(&self) -> Result<Option<usize>> {
		self.base.base.rpc_ws_max_connections()
	}

	fn rpc_cors(&self, is_dev: bool) -> Result<Option<Vec<String>>> {
		self.base.base.rpc_cors(is_dev)
	}

	fn default_heap_pages(&self) -> Result<Option<u64>> {
		self.base.base.default_heap_pages()
	}

	fn force_authoring(&self) -> Result<bool> {
		self.base.base.force_authoring()
	}

	fn disable_grandpa(&self) -> Result<bool> {
		self.base.base.disable_grandpa()
	}

	fn max_runtime_instances(&self) -> Result<Option<usize>> {
		self.base.base.max_runtime_instances()
	}

	fn announce_block(&self) -> Result<bool> {
		self.base.base.announce_block()
	}

	fn telemetry_endpoints(
		&self,
		chain_spec: &Box<dyn sc_service::ChainSpec>,
	) -> Result<Option<sc_telemetry::TelemetryEndpoints>> {
		self.base.base.telemetry_endpoints(chain_spec)
	}
}
